File: apps/android/app/src/main/java/com/sennnen/mav/ui/aura/AuraMlSignalsCard.kt
package com.sennnen.mav.ui.aura

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Air
import androidx.compose.material.icons.filled.DirectionsRun
import androidx.compose.material.icons.filled.Memory
import androidx.compose.material.icons.filled.Psychology
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import com.sennnen.mav.ui.AppViewModel
import kotlin.math.roundToInt

// Android twin of iOS Strand/UI/AuraMLSignalsCard.swift — the read-only surface for the on-device ML
// layer (MlEngine). Shows only the signals that have a value yet, so it stays quiet until the strap has
// streamed enough beats. Honestly labelled as an on-device estimate, never a medical reading. Surfaces
// stress, cardio fitness and respiration; excludes heart-rhythm (AFib) + blood pressure, matching the
// engine's published set. Pure Material 3.
@Composable
fun AuraMlSignalsCard(vm: AppViewModel) {
    val p = Aura.palette
    val engine = vm.mlEngine
    val backboneActive by engine.backboneActive.collectAsStateWithLifecycle()
    val stressLoad by engine.stressLoad.collectAsStateWithLifecycle()
    val vo2max by engine.vo2max.collectAsStateWithLifecycle()
    val respiration by engine.respirationRate.collectAsStateWithLifecycle()

    val hasAny = stressLoad != null || vo2max != null || respiration != null

    Column(verticalArrangement = Arrangement.spacedBy(14.dp)) {
        AuraSectionHeader(title = "On-device signals")

        if (backboneActive) {
            Row(
                Modifier.fillMaxWidth().padding(horizontal = 4.dp),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Icon(Icons.Filled.Memory, null, tint = p.good, modifier = Modifier.size(14.dp))
                Text(
                    "Pulse-PPG model active on device",
                    style = AuraType.caption, color = p.ink.copy(alpha = 0.6f),
                )
            }
        }

        if (!hasAny) {
            AuraDarkCard(padding = 18.dp) {
                Text(
                    "Wear your strap for a few minutes. Stress load, cardio fitness and respiration are " +
                        "estimated on device, privately.",
                    style = AuraType.sub, color = p.ink.copy(alpha = 0.6f),
                )
            }
        } else {
            AuraDarkCard(padding = 0.dp) {
                Column(Modifier.padding(vertical = 4.dp)) {
                    var shown = false
                    stressLoad?.let {
                        StressRow(it)
                        shown = true
                    }
                    vo2max?.let {
                        if (shown) androidx.compose.material3.HorizontalDivider(
                            color = p.hairline, modifier = Modifier.padding(start = 18.dp),
                        )
                        MlRow(Icons.Filled.DirectionsRun, "Cardio fitness", it.roundToInt().toString(), "VO₂max", AuraFamily.CHARGE.glow)
                        shown = true
                    }
                    respiration?.let {
                        if (shown) androidx.compose.material3.HorizontalDivider(
                            color = p.hairline, modifier = Modifier.padding(start = 18.dp),
                        )
                        MlRow(Icons.Filled.Air, "Respiration", it.roundToInt().toString(), "br/min", AuraFamily.HEART.glow)
                    }
                }
            }
            Text(
                "Estimated on your phone from the strap's optical signal. A screen, not a diagnosis.",
                style = AuraType.caption, color = p.ink.copy(alpha = 0.4f),
                modifier = Modifier.padding(horizontal = 4.dp),
            )
        }
    }
}

// Two-line stress row (label + qualitative word) with a big load number on the right — mirrors the iOS
// stressRow. Distinct from MlRow because stress carries a subtitle and no trailing unit.
@Composable
private fun StressRow(load: Double) {
    val p = Aura.palette
    val word = if (load >= 66) "High" else if (load >= 40) "Moderate" else "Low"
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 18.dp, vertical = 13.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Icon(Icons.Filled.Psychology, null, tint = AuraFamily.EFFORT.glow, modifier = Modifier.size(20.dp))
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
            Text("Stress load", style = AuraType.label, color = p.ink.copy(alpha = 0.92f))
            Text("$word · from HRV & heart-rate", style = AuraType.caption, color = p.ink.copy(alpha = 0.55f))
        }
        Text(load.roundToInt().toString(), style = AuraType.number(22.sp), color = p.ink, fontWeight = FontWeight.Normal)
    }
}

@Composable
private fun MlRow(icon: androidx.compose.ui.graphics.vector.ImageVector, title: String, value: String, unit: String, tint: androidx.compose.ui.graphics.Color) {
    val p = Aura.palette
    Row(
        Modifier.fillMaxWidth().padding(horizontal = 18.dp, vertical = 13.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
    ) {
        Icon(icon, null, tint = tint, modifier = Modifier.size(20.dp))
        Text(title, style = AuraType.label, color = p.ink.copy(alpha = 0.92f))
        Spacer(Modifier.weight(1f))
        Row(verticalAlignment = Alignment.Bottom, horizontalArrangement = Arrangement.spacedBy(4.dp)) {
            Text(value, style = AuraType.number(22.sp), color = p.ink, fontWeight = FontWeight.Normal)
            Text(unit, style = AuraType.caption, color = p.ink.copy(alpha = 0.5f), modifier = Modifier.padding(bottom = 3.dp))
        }
    }
}

File: apps/ios/Maverick/UI/AuraMLSignalsCard.swift
import SwiftUI

// AuraMLSignalsCard — the read-only surface for the on-device ML layer (StrandML). Observes the engine
// DIRECTLY (a nested ObservableObject doesn't republish through AppModel) and shows only the signals that
// have a value yet, so it stays quiet until the strap has streamed enough beats. Every line is honestly
// labelled as an on-device estimate / screen, never a medical reading.

struct AuraMLSignalsCard: View {
  @ObservedObject var engine: StrandMLEngine

  var body: some View {
    VStack(alignment: .leading, spacing: 14) {
      AuraSectionHeader(title: "On-device signals")

      if engine.backboneActive {
        HStack(spacing: 8) {
          Image(systemName: "cpu")
            .font(.system(size: 12, weight: .semibold)).foregroundStyle(AuraDesign.good)
          Text("Pulse-PPG model active on device")
            .font(AuraDesign.caption).foregroundStyle(AuraDesign.ink.opacity(0.6))
          Spacer()
        }
        .padding(.horizontal, 4)
      }

      if !hasAny {
        Text("Wear your strap for a few minutes. Cardiac rhythm, stress load and cardio fitness are estimated on device, privately.")
          .font(AuraDesign.sub).foregroundStyle(AuraDesign.ink.opacity(0.6))
          .fixedSize(horizontal: false, vertical: true)
          .auraDarkCard(padding: 18)
      } else {
        VStack(spacing: 0) {
          if let afib = engine.afib, afib.reliable { rhythmRow(afib); divider }
          if let load = engine.stressLoad { stressRow(load); divider }
          if let vo2 = engine.vo2max { row("Cardio fitness", String(format: "%.0f", vo2), "VO₂max", .charge); divider }
          if let br = engine.respirationRate { row("Respiration", String(format: "%.0f", br), "br/min", .vitals) }
        }
        .padding(.vertical, 4)
        .background(AuraDesign.card, in: AuraDesign.tileShape)
        .overlay(AuraDesign.tileShape.strokeBorder(AuraDesign.hairline, lineWidth: 1))

        Text("Estimated on your phone from the strap's optical signal. A screen, not a diagnosis.")
          .font(AuraDesign.caption).foregroundStyle(AuraDesign.ink.opacity(0.4))
          .fixedSize(horizontal: false, vertical: true)
          .padding(.horizontal, 4)
      }
    }
  }

  private var hasAny: Bool {
    engine.afib?.reliable == true || engine.stressLoad != nil
      || engine.vo2max != nil || engine.respirationRate != nil
  }

  @ViewBuilder private var divider: some View {
    Rectangle().fill(AuraDesign.ink.opacity(0.08)).frame(height: 1).padding(.leading, 18)
  }

  private func rhythmRow(_ afib: AFibScreener.AFibResult) -> some View {
    let irregular = afib.irregular
    return HStack(spacing: 14) {
      Image(systemName: irregular ? "waveform.path.ecg" : "heart")
        .font(.system(size: 16, weight: .medium))
        .foregroundStyle(irregular ? AuraDesign.fair : AuraDesign.good)
        .frame(width: 26)
      VStack(alignment: .leading, spacing: 2) {
        Text("Heart rhythm").font(AuraDesign.label).foregroundStyle(AuraDesign.ink.opacity(0.92))
        Text(irregular ? "Irregular rhythm noticed" : "Looks regular")
          .font(AuraDesign.caption).foregroundStyle(AuraDesign.ink.opacity(0.55))
      }
      Spacer()
      AuraStatusChip(text: irregular ? "Check" : "Regular",
                     kind: irregular ? .caution : .positive)
    }
    .padding(.horizontal, 18).padding(.vertical, 13)
  }

  private func stressRow(_ load: Double) -> some View {
    let word: String = load >= 66 ? "High" : load >= 40 ? "Moderate" : "Low"
    return HStack(spacing: 14) {
      Image(systemName: "brain.head.profile")
        .font(.system(size: 16, weight: .medium))
        .foregroundStyle(AuraDesign.Family.effort.glow)
        .frame(width: 26)
      VStack(alignment: .leading, spacing: 2) {
        Text("Stress load").font(AuraDesign.label).foregroundStyle(AuraDesign.ink.opacity(0.92))
        Text("\(word) · from HRV & heart-rate")
          .font(AuraDesign.caption).foregroundStyle(AuraDesign.ink.opacity(0.55))
      }
      Spacer()
      Text("\(Int(load.rounded()))")
        .font(AuraDesign.number(22)).foregroundStyle(AuraDesign.ink).monospacedDigit()
    }
    .padding(.horizontal, 18).padding(.vertical, 13)
  }

  private func row(_ title: String, _ value: String, _ unit: String,
                   _ family: AuraDesign.Family) -> some View {
    HStack(spacing: 14) {
      Image(systemName: "figure.run.circle")
        .font(.system(size: 16, weight: .medium))
        .foregroundStyle(family.glow)
        .frame(width: 26)
      Text(title).font(AuraDesign.label).foregroundStyle(AuraDesign.ink.opacity(0.92))
      Spacer()
      HStack(alignment: .firstTextBaseline, spacing: 4) {
        Text(value).font(AuraDesign.number(22)).foregroundStyle(AuraDesign.ink).monospacedDigit()
        Text(unit).font(AuraDesign.caption).foregroundStyle(AuraDesign.ink.opacity(0.5))
      }
    }
    .padding(.horizontal, 18).padding(.vertical, 13)
  }
}
